timemachine/src/time_travel.rs

```diff
@@ -0,0 +1,742 @@
+            };
+            write_redo_stmt.execute(rusqlite::params![
+              &entry.versionstamp,
+              range.timestamp_ms,
+              &entry.key,
+              &value,
+              entry.value_encoding,
+              &entry.versionstamp[0..10],
+            ])?;
+          }
+          last_key = Some(range_entry.key);
+        }
+      }
+      let last_key = last_key.as_ref().unwrap();
+      update_redo_cursor(&tx, last_key)?;
+      tx.commit()?;
+    }
+
+    Ok(last_key)
+  }
+}
+
+fn query_redo_cursor(
+  conn: &rusqlite::Connection,
+) -> Result<Option<MutationRangeKey>> {
+  conn.query_row(
+      "select format_version, monoseq, first_versionstamp12, last_versionstamp12 from tt_redo_cursor where zero = 0",
+      [],
+      |row| Ok(MutationRangeKey{
+        format_version: row.get(0)?,
+        monoseq: row.get(1)?,
+        first_versionstamp12: row.get(2)?,
+        last_versionstamp12: row.get(3)?,
+      })).optional().map_err(anyhow::Error::from)
+}
+
+fn update_redo_cursor(
+  conn: &rusqlite::Connection,
+  key: &MutationRangeKey,
+) -> Result<()> {
+  conn.execute(
+    "replace into tt_redo_cursor (zero, format_version, monoseq, first_versionstamp12, last_versionstamp12) values (0, ?, ?, ?, ?)",
+    rusqlite::params![&key.format_version, &key.monoseq, &key.first_versionstamp12, &key.last_versionstamp12],
+  )?;
+  Ok(())
+}
+
+fn get_config(
+  conn: &rusqlite::Connection,
+  key: &str,
+) -> Result<Option<String>> {
+  conn
+    .query_row("select v from tt_config where k = ?", [key], |row| {
+      row.get(0)
+    })
+    .optional()
+    .map_err(anyhow::Error::from)
+}
+
+fn set_config(
+  conn: &rusqlite::Connection,
+  key: &str,
+  value: &str,
+) -> Result<()> {
+  conn.execute("replace into tt_config (k, v) values (?, ?)", [key, value])?;
+  Ok(())
+}
+
+struct InitialSnapshotRange {
+  key: SnapshotRangeKey,
+  pulled: bool,
+}
+
+fn list_initial_snapshot_ranges(
+  conn: &rusqlite::Connection,
+) -> Result<Vec<InitialSnapshotRange>> {
+  conn
+    .prepare_cached(
+      "select format_version, monoseq, seq, pulled from tt_initial_snapshot_ranges order by format_version, monoseq, seq",
+    )?
+    .query_map([], |row| {
+      Ok(InitialSnapshotRange {
+        key: SnapshotRangeKey {
+          format_version: row.get(0)?,
+          monoseq: row.get(1)?,
+          seq: row.get(2)?,
+        },
+        pulled: row.get::<_, u64>(3)? != 0,
+      })
+    })?
+    .collect::<Result<Vec<_>, rusqlite::Error>>()
+    .map_err(anyhow::Error::from)
+}
+
+fn insert_initial_snapshot_ranges(
+  conn: &rusqlite::Connection,
+  ranges: &[SnapshotRangeKey],
+) -> Result<()> {
+  for range in ranges {
+    conn.execute(
+      "insert into tt_initial_snapshot_ranges (format_version, monoseq, seq, pulled) values (?, ?, ?, ?)",
+      rusqlite::params![&range.format_version, &range.monoseq, &range.seq, &0u64],
+    )?;
+  }
+  Ok(())
+}
+
+fn replay_redo_or_undo_in_tx(
```

Comment:
Contrary to the name, you don't seem to be using a transaction here.

Reply:
`in_tx` means this method should be called within a transaction.

Reply:
Why not just pass the `tx` to this method then? That makes this much clearer.